Exponential back-off support for Google PaLM api

mtronci opened the pull request:
This PR adds exponential back-off to the Google PaLM api to gracefully handle rate limiting errors.

mtronci committed:
Adedd exponential back-off for API calls

mtronci committed:
Adedd exponential back-off for API calls, with print statements

vowelparrot commented on langchain/embeddings/google_palm.py:
```diff
@@ -6,6 +6,38 @@
+                    return f(*args, **kwargs)
+                except ExceptionToCheck:
+                    msg = "%s, Retrying in %d seconds..." % (str(ExceptionToCheck), mdelay)
+                    print(msg)
```
Should use `logger.info()` instead

mtronci replied:
many thanks for reviewing.
I'll look to replicate a similar approach.

vowelparrot reviewed (commented):
Would be preferable to reuse / share / extend the `completion_with_retry` decorators that use `tenacity`. You can reference these at `langchain/{chat_models|embeddings|llms}/openai.py` and in the confluence dataloader

mtronci committed:
Implemented PaLM retries using tenacity

mtronci committed:
Latest poetry.lock

mtronci commented:
I have reworked the changes, and also added retry support to the PaLM chat.
I have gone for a simple approach, it is not exactly the same as the OpenAI one.
The decorator is stored in a class attribute, and then used to decorate the relevant methods.

dev2049 committed:
fmt

dev2049 committed:
fmt

dev2049 committed:
nit

dev2049 committed:
cr

dev2049 committed:
cr

dev2049 committed:
lint

brotchie commented:
Thanks so much for adding this @mtronci and to all reviewers 🙏

mtronci commented:
thanks all for reviewing and merging!
